File: PaperLib/Utilities/Preference.swift
//
//  Preference.swift
//  PaperLib
//
//  Created by GeoffreyChen on 19/02/2022.
//

import Foundation

extension UserDefaults {
    @objc var preferColorTheme: String {
        get {
            return string(forKey: "preferColorTheme") ?? "System Default"
        }
        set {
            set(newValue, forKey: "preferColorTheme")
        }
    }

    @objc var invertColor: Bool {
        get {
            return bool(forKey: "invertColor")
        }
        set {
            set(newValue, forKey: "invertColor")
        }
    }
}

func registerPreference () {
    UserDefaults.standard.register(
        defaults:
            [
                "appLibFolder": FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].appendingPathComponent("paperlib").absoluteString,
                "invertColor": false,
                "preferColorTheme": "System Default",
                "deleteSourceFile": true,
                "showSidebarCount": true,
                "allowRoutineMatch": false,
                "rematchInterval": 7,
                "syncAPIKey": "",
                "useSync": false,
                "exportReplacement": Data.init(),
                "enableExportReplacement": false,

                // Scraper
                "pdfBuiltinScraper": true,
                "arXivScraper": true,
                "doiScraper": true,
                "teScraper": true,
                "ieeeScraper": true,
                "dblpScraper": true,
                "cvfScraper": true,

                // ScraperAPIKey
                "ieeeAPIKey": ""
            ]
    )
}

File: src/utils/preference.js
import Store from "electron-store";
import path from "path";
import os from "os";
import moment from "moment";

const defaultPreferences = {
    appLibFolder: path.join(os.homedir(), "Documents", "paperlib"),
    deleteSourceFile: false,
    enableExportReplacement: true,
    exportReplacement: [],
    useSync: false,
    syncAPIKey: "",
    allowRoutineMatch: true,
    lastRematchTime: moment().unix(),
    rematchInterval: 7,

    pdfBuiltinScraper: true,
    arXivScraper: true,
    doiScraper: true,
    teScraper: true,
    dblpScraper: true,
    cvfScraper: true,
    ieeeScraper: true,
    ieeeAPIKey: ""
}


export class Preference {
    constructor() {
        this.store = new Store();

        // Initialize default preferences
        for (let key in defaultPreferences) {
            if (!this.store.has(key)) {
                this.store.set(key, defaultPreferences[key]);
            }
        }
    }

    get(key) {
        if (this.store.has(key)) {
            return this.store.get(key);
        } else {
            this.set(key, defaultPreferences[key]);
            return defaultPreferences[key];
        }
    }

    set(key, value) {
        console.log(key, value)
        this.store.set(key, value);
    }
}
